src/trie.js

```diff
@@ -1,102 +1,518 @@
-const _ = require('lodash');
-const {scopeCompare, normalizeScopeSet, mergeScopeSets} = require('taskcluster-lib-scopes');
+const assert = require('assert');
+const ScopeSetBuilder = require('./scopesetbuilder');
 
-let normalized = (scopes) => {
-  scopes.sort(scopeCompare);
-  return normalizeScopeSet(scopes);
+// Construct character SUBSTITUTE used for representation of <..>
+const PARAM = '\u001a';
+const PARAM_TO_END = /\u001a.*$/;
+
+/**
+ * Transform rules from {pattern, scopes} to {pattern, scopes, matched, paramed},
+ * where:
+ *  - matched, is the non-parameterized scopes, and
+ *  - paramed, is the parameterized scopes with PARAM instead of '<..>'
+ */
+const transformRules = (rules) => rules.map(({pattern, scopes}) => {
+  const hasKleene = pattern.endsWith('*');
+  let matched = [];
+  let paramed = [];
+  for (const scope of scopes) {
+    if (hasKleene && scope.includes('<..>')) {
+      const s = scope.replace('<..>', PARAM);
+      paramed.push(s);
+      // We forbid more than one parameter injection
+      if (s.includes('<..>')) {
+        const err = new Error(`parameterized scope '${scope}' contains multiple '<..>'`);
+        err.code = 'InvalidScopeError';
+        err.scope = scope;
+        throw err;
+      }
+      // We forbid ambiguous kleenes
+      if (s.endsWith('*' + PARAM)) {
+        const err = new Error(`parameterized scope '${scope}' ends with '*<..>' which implies an ambiguous kleene`);
+        err.code = 'InvalidScopeError';
+        err.scope = scope;
+        throw err;
+      }
+    } else {
+      matched.push(scope);
+    }
+  }
+  matched = ScopeSetBuilder.normalizeScopeSet(matched);
+  paramed = ScopeSetBuilder.normalizeScopeSet(paramed);
+  return {pattern, scopes, matched, paramed};
+});
+
+/**
+ * Given a sorted list of scopes that contains PARAM, replace PARAM with param
+ * and return the sorted result.
+ */
+const withParam = (scopes, param) => {
+  const pattern = param.endsWith('*') ? PARAM_TO_END : PARAM;
+  scopes = scopes.map(s => s.replace(pattern, param));
+  return ScopeSetBuilder.normalizeScopeSet(scopes);
+};
+
+/** Node in a trie */
+class Node {
+  constructor(end = [], enter = [], paramed = [], kleeneOnly = []) {
+    this.end = end;               // scopes given if input ends at this node
+    this.enter = enter;           // scopes given if input traverses through this node
+    this.paramed = paramed;       // scopes given if input traverses through this node with parameter
+    this.kleeneOnly = kleeneOnly; // scopes given in addition to end, enter, paramed if input ends with kleene
+    this._kleeneCache = null;     // scopes given if input ends with kleene at this node (lazily computed)
+    this.children = new Map();    // map from next character to child node
+  }
+
+  dump() {
+    return {
+      end: this.end,
+      enter: ScopeSetBuilder.normalizeScopeSet([...this.enter, ...withParam(this.paramed, '<..>')]),
```

Comment:
Not that it matters outside of tests, but why is paramed not listed separately here?

Reply:
I actually meant to remove this... dump() was exclusively for debugging, we can keep it :)

I just wanted to keep a minimalist representation of the nodes...